README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# snahelper <img src="man/figures/logo.png" align="right"/>

[![R-CMD-check](https://github.com/schochastics/snahelper/workflows/R-CMD-check/badge.svg)](https://github.com/schochastics/snahelper/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/snahelper)](https://cran.r-project.org/package=snahelper)
[![Downloads](https://cranlogs.r-pkg.org/badges/snahelper)](https://CRAN.R-project.org/package=snahelper)

`snahelper` provides a set RStudio Addin for social network analysis.
The main addin is the `SNAhelper` which provides a simple GUI to do
common network analytic tasks and visualize a network with
[ggraph](https://ggraph.data-imaginist.com/).

The second addin, called `Netbuilder` allows you to quickly build small
networks with a small “canvas” to draw on. The network can be exported
as an igraph object at the end of the session by clicking on “Done”.

The third addin `Netreader` is meant to facilitated the import of raw
network data. It provides a GUI to easily read network and attribute
data and combine them to an igraph object. The underlying code of the
import procedure is shown at the end. This should help users to learn
importing data themselves.

The fourth addin `Componentlayouter` allows to layout networks with
several components manually by placing them on an empty canvas.
Components can also be rotated and resized. After finishing the session,
the layout is saved as x and y vertex attributes.

## Quick look at the SNAhelper

![](man/figures/snahelper.gif)

Code to reproduce the used network.

``` r
library(tidyverse)
library(igraph)
url <- "https://raw.githubusercontent.com/evelinag/StarWars-social-network/master/networks/starwars-full-interactions-allCharacters.json"
df <- jsonlite::fromJSON(url)

df$nodes$id <- 1:nrow(df$nodes)
df$nodes$display <- df$nodes$name
df$nodes$name <- df$nodes$id-1
g <- graph_from_data_frame(df$links,directed = F,vertices = df$nodes)

V(g)$name <- V(g)$display

g <- remove.vertex.attribute(g,"id")
g <- remove.vertex.attribute(g,"display")
g <- delete.vertices(g,which(degree(g)==0))

V(g)$display <- ifelse(V(g)$value>75,V(g)$name,"")
V(g)$colour <- ifelse(V(g)$display=="",NA,V(g)$display)

g
```

## Quick look at the Netreader

`Netreader` should be pretty selfexplanatory. The first two tabs allow
you to import raw data (edges and attributes). Make sure to specify file
delimiters, etc. according to the shown preview.

![](man/figures/Netreader1.png)

Using the `Netreader` should comes with a learning effect (hopefully).
The last tab shows the R code to produce the network with the chosen
data **without** using the Addin. ![](man/figures/Netreader2.png)

The network will be saved in your global environment once you click
“Done”.

## Quick look at the Netbuilder

![](man/figures/Netbuilder.gif)

## Quick look at the Componentlayouter

![](man/figures/Componentlayouter.gif)

Highlight an igraph object in your script and selcet the
`Componentlayouter` from the RStudio addin menu. Components are added by
clicking on the canvas and can be resized/rotated and re-positioned
using the buttons at the bottom. The final layout is stored as x and y
coordinates in the igraph object.

## Installation

``` r
# developer version
#install.packages(remotes)
remotes::install_github("schochastics/snahelper")

#CRAN version
install.packages("snahelper")
```

To work properly, you also need
[graphlayouts](https://github.com/schochastics/graphlayouts), which adds
new layout algorithms.

``` r
install.packages("graphlayouts")
```

## Usage of SNAhelper

To use the main addin, simply highlight a network in your script and
select `SNAhelper` from the Addin dropdown menu.

## Interface

### Layout

The layout tab allows you to choose from all implemented algorithms in
`igraph` and some layouts from `graphlayouts`. The default is a stress
based layout and also the recommended choice. See [my
blog](http://blog.schochastics.net/post/stress-based-graph-layouts/) for
an explanation. In the tweak section you can move individual nodes
around. Choose the node from the dropdown menu and click on its new
position in the plot.

![](man/figures/layout.png)

### Node Attribute Manager

The Node Attribute Manager shows all existing node attributes in a
sortable table. In addition, you can calculate some new ones (centrality
and clustering). The functions automatically choose the right version of
indices, depending if the network is
directed/weighted/undirected/unweighted.

![](man/figures/NattributeMan.png)

### Nodes

This is where you can style your nodes. You can either do it manually,
by choosing a color/size for all nodes together, or based on an
attribute.

![](man/figures/nodes.png)

### Edge Attribute Manager

Same as Node Attribute Manager but for edges. So far only shows existing
edge attributes.

![](man/figures/EattributeMan.png)

### Edges

You can style your edges here. `snahelper` automatically detects if your
network is directed and adds arrows if the network is directed. The
other options are similar to the nodes tab. The snahelper automatically
chooses the appropriate edge geom. If multiple edges are present, it
uses `geom_edge_parallel0()`. Otherwise `geom_edge_link0()`.

![](man/figures/edges.png)

### Result

The result tab shows the network in its full size. If you are satisfied
with the results, hit the Done button and the R code to produce the plot
is automatically inserted in your script. Or you can directly save the
result as a png file.

![](man/figures/result.png)


NEWS.md
# snahelper 1.4.2

* partial code refactoring and new code style

# snahelper 1.4.1

* fixed bug for directed degree in snahelper
* fixed weighted adjacency matrix bug in netreader

# snahelper 1.4.0

* `Snahelper` is now more reproducible by also returning the coordinates explicitly
* fixed deprecated FA 5 sliders name (changed to sliders-h)

# snahelper 1.3.0

* added layout via stored node variables x and y 
* added `drop = FALSE` to `Netreader` for cases where only one attribute is read
* minor bug fixes

# snahelper 1.2.0

* added `componentlayouter`

# snahelper 1.1.0

* Fixed a bug that prevented the calculation of in/out degree in directed networks  
* Fixed a bug that produced warnings for directed networks
* Fixed label of variable name field for `netbuilder` and `netreader`
* Switched plot and control panel layout to remove scrolling issue. 
* Use {{DT}} for table rendering due to persistent issue with character attributes.

# snahelper 1.0.0

* **BREAKING** only works with the newest ggraph version (2.0.0) (error message added)
* added `Netbuilder` addin
* added `Netreader` addin
* edge geom is chosen automatically between `geom_edge_link0()` and `geom_edge_parallel0()`
* discrete edge colors are supported now
* added support for `layout_with_focus()` and `layout_with_centrality()`

# snahelper 0.3.0

* tweaking node position is now possible via clicking on desired location
* added `formatR` for nicer code output
* added backbone layout support
* centrality indices in node attribute manager automatically adapt to type of graph (directed/weighted/etc.)

# snahelper 0.2.0

* added option to repel node labels
* removed curvature for edges and added manual alpha tuning
* switched from `geom_edge_arc` to `geom_edge_fan`
* changed some default colors
* increased max node size
* added `colourpicker` support

# snahelper 0.1.0

* initial version


R/helper.R
headingOutput <- function(heading, height = "20px", css = "color: #FF7F00; text-decoration: underline;") {
    fillCol(tags$div(style = css, strong(heading)), height = height)
}

is.validColour <- function(x) {
    if (is.null(x)) {
        return(TRUE)
    } else if (x %in% c(colours.available, "NA", "NULL")) {
        return(TRUE)
    } else if (grepl("#[0-9a-fA-F]{6}$", x)) {
        return(TRUE)
    } else {
        return(FALSE)
    }
}

create_attribute_df <- function(g, which = "nodes") {
    if (which == "nodes") {
        vattrs <- igraph::vertex_attr_names(g)
        df_attrs <- data.frame(id = seq_len(igraph::vcount(g)))
        if (length(vattrs) > 0) {
            for (i in seq_along(vattrs)) {
                df_attrs[[vattrs[i]]] <- igraph::get.vertex.attribute(g, vattrs[i])
            }
        }
    } else if (which == "edges") {
        eattrs <- igraph::edge_attr_names(g)
        df_attrs <- data.frame(id = seq_len(igraph::ecount(g)))
        if (length(eattrs) > 0) {
            for (i in seq_along(eattrs)) {
                df_attrs[[eattrs[i]]] <- igraph::get.edge.attribute(g, eattrs[i])
            }
        }
    }
    df_attrs
}

# JS code ----
jscodeHeight <-
    '$(document).on("shiny:connected", function(e) {
    var jsHeight = document.documentElement.clientHeight;
    Shiny.onInputChange("ViewerHeight",jsHeight);
  });'

jscodeWidth <-
    '$(document).on("shiny:connected", function(e) {
    var jsWidth = document.documentElement.clientWidth;
    Shiny.onInputChange("ViewerWidth",jsWidth);
  });'

jsColourSelector <- I(
    '{
  option: function(item, escape) {
  return "<div><div style=\'width:25px; height:15px; background-color:" + item.rgb + "; float:left; vertical-align:bottom\'></div>&nbsp;" + escape(item.name) + "</div>";
  }
  }'
)


closest_node <- function(p, xy) {
    if (nrow(xy) == 0) {
        return(-1)
    }
    dists <- sqrt((p[1] - xy[, 1])^2 + (p[2] - xy[, 2])^2)
    mdist <- min(dists)
    if (mdist < 0.5) {
        return(which.min(dists))
    } else {
        return(-1)
    }
}


R/constants.R
input.width <- "50%"
input.width2 <- "90%"
line.height <- "70px"
heading.height <- "30px"

layouts.available <- list(
    "graphlayouts" = c(
        "Stress" = "graphlayouts::layout_with_stress",
        "Backbone" = "graphlayouts::layout_as_backbone",
        "Radial focus" = "graphlayouts::layout_with_focus",
        "Radial centrality" = "graphlayouts::layout_with_centrality"
    ),
    "igraph" = c(
        "Circle" = "layout_in_circle", "Nicely" = "layout_nicely",
        "Random" = "layout_randomly",
        "Fruchterman-Reingold" = "layout_with_fr",
        "MDS" = "layout_with_mds",
        "Sugiyama" = "layout_with_sugiyama",
        "drl" = "layout_with_drl",
        "gem" = "layout_with_gem",
        "graphopt" = "layout_with_graphopt",
        "kk" = "layout_with_kk",
        "lgl" = "layout_with_lgl"
    )
)

colours.available <- c(colors()[!grepl("grey", colors())])
fonts.available <- names(pdfFonts())

AttrNameImport <- paste0(
    "vnames <- get.vertex.attribute(g,'name')\n",
    "identCol <- which(apply(attrs,2,function(x) all(x%in%vnames)))[1]\n",
    "anames <- attrs[,identCol]\n",
    "attrs <- attrs[,-identCol,drop = FALSE]\n",
    "perm <- match(vnames,anames)\n",
    "for(attr in names(attrs)){\n",
    "   g <- set_vertex_attr(g,name = attr,value = attr[[attr]][perm])\n",
    "}\n"
)

AttrRowImport <- paste0(
    "for(attr in names(attrs)){\n",
    "   g <- set_vertex_attr(g,name = attr,value = attrs[[attr]])\n",
    "}\n"
)


R/Netbuilder.R
#' Netbuilder
#'
#' \code{Netbuilder} is an RStudio-Addin that allows you to create small networks with point and click.
#'
#' @details To run the addin, select \code{Netbuilder} from the Addins-menu within RStudio.
#' @return \code{Netbuilder} returns the created network as igraph object.
#' @import miniUI
#' @import shiny
#' @import ggplot2
#' @import ggraph
#' @import rstudioapi
#' @import igraph
#' @name Netbuilder
NULL

Netbuilder <- function() {
    rv <- reactiveValues(
        g = igraph::graph.empty(n = 0, directed = FALSE), xy = matrix(0, 0, 2),
        start = -1, end = -1, el = matrix(0, 0, 2)
    )
    # ui ----
    ui <- miniPage(
        tags$script(jscodeWidth),
        tags$script(jscodeHeight),
        tags$style(type = "text/css", ".selectize-dropdown{ width: 200px !important; }"),
        tags$style(type = "text/css", ".form-group.shiny-input-container{width:50%;}"),
        tags$style(type = "text/css", "#makeD {background-color: rgba(30,144,255,1); color: white}"),
        tags$style(type = "text/css", "#makeU {background-color: rgba(30,144,255,1); color: white}"),
        tags$style(type = "text/css", "#clearG {background-color: rgba(220,20,60,1); color: white}"),
        gadgetTitleBar("Netbuilder"),
        plotOutput("Graph1", width = "80%", height = "80%", click = "add_edge", dblclick = "add_vertex"),
        # hover = hoverOpts(id = "showXY",delay = 500,delayType = "throttle")),
        fillRow(
            height = line.height, width = "100%",
            actionButton("makeU", "make undirected"),
            actionButton("makeD", "make directed"),
            actionButton("clearG", "clear"),
            textAreaInput("text", label = "", value = "", placeholder = "enter name", height = "35px")
        ),
        p(
            strong("Double Click:"), " add vertex.",
            strong("Click on two nodes:"), " add edge.",
            strong("Done:"), "Export graph to global enivronment with name from text input field"
        )
    )

    # server ----
    server <- function(input, output, session) {
        #--------------------#
        # observe click ----
        #--------------------#
        shiny::observeEvent(input$add_edge, {
            pts <- c(input$add_edge$x, input$add_edge$y)
            g <- rv$g
            xy <- rv$xy
            start <- rv$start
            end <- rv$end
            closest_nID <- closest_node(pts, xy)
            if (start == -1) {
                start <- closest_nID
                rv$start <- start
            } else {
                end <- closest_nID
                if (end != -1) {
                    g <- igraph::add.edges(g, c(start, end))
                    rv$start <- -1
                    rv$end <- -1
                    rv$g <- g
                    rv$el <- rbind(rv$el, c(start, end))
                } else {
                    g <- igraph::add.vertices(g, 1)
                    rv$start <- -1
                    rv$end <- -1
                    xy <- rv$xy
                    xy <- rbind(xy, pts)
                    g <- igraph::add.edges(g, c(start, nrow(xy)))
                    rv$xy <- xy
                    rv$g <- g
                    rv$el <- rbind(rv$el, c(start, nrow(xy)))
                }
            }
            gg_reactive()
        })
        #--------------------#
        # observe dbclick ----
        #--------------------#
        shiny::observeEvent(input$add_vertex, {
            g <- rv$g
            g <- igraph::add.vertices(g, 1)
            rv$g <- g
            xy <- rv$xy
            xy <- rbind(xy, c(input$add_vertex$x, input$add_vertex$y))
            rv$xy <- xy
            rv$start <- -1
            gg_reactive()
        })
        #--------------------#
        # observe dbclick ----
        #--------------------#
        # output$coordXY <- renderPrint({
        #   cat("Hover:\n")
        #   str(input$showXY)
        # })
        #--------------------#
        # make (un)directed ----
        #--------------------#
        shiny::observeEvent(input$makeD, {
            if (igraph::ecount(rv$g) != 0) {
                # el <- igraph::get.edgelist(rv$g)
                el <- rv$el
                g <- igraph::graph_from_edgelist(el, directed = TRUE)
                rv$g <- g
            }
        })
        shiny::observeEvent(input$makeU, {
            if (igraph::ecount(rv$g) != 0) {
                # el <- igraph::get.edgelist(rv$g)
                el <- rv$el
                g <- igraph::graph_from_edgelist(el, directed = FALSE)
                rv$g <- g
            }
        })

        shiny::observeEvent(input$clearG, {
            g <- igraph::graph.empty(n = 0, directed = FALSE)
            rv$g <- g
            xy <- matrix(0, 0, 2)
            rv$xy <- xy
            rv$el <- xy
        })
        #-------------------#
        # plot ----
        #-------------------#
        gg_reactive <- reactive({
            code_layout <- "ggraph(rv$g,layout = \"manual\", x = rv$xy[,1], y = rv$xy[,2])"
            code_nodes <- "geom_node_point(shape = 21,fill = \"grey25\",size=8)"
            code_label <- "geom_node_text(label=1:nrow(rv$xy),size=6,col=\"white\")"
            if (igraph::is.directed(rv$g)) {
                code_arrow <- paste0(
                    ",\narrow = arrow(angle = 25, length = unit(0.15, \"inches\")",
                    ",\nends = \"last\", type = \"closed\")",
                    ",\nend_cap = circle(", 8, ", \"pt\"))"
                )
                code_edges <- "geom_edge_link(edge_width=0.4,edge_colour=\"grey66\""

                code_edges <- paste0(code_edges, ", ", code_arrow)
            } else {
                code_edges <- "geom_edge_link0(edge_width=0.4,edge_colour=\"grey66\")"
            }
            code_theme <- "theme_graph(foreground=\"black\",border=TRUE)"
            code_scale <- "scale_x_continuous(limits=c(0,10)) + scale_y_continuous(limits=c(0,10))"
            code <- paste(code_layout, code_edges, code_nodes, code_label, code_theme, code_scale, sep = " + ")
            p <- code

            return(p)
        })

        # render plot
        ggnet <- renderPlot({
            eval(parse(text = gg_reactive()))
        })
        output$Graph1 <- ggnet

        # cancel ----
        observeEvent(input$cancel, {
            invisible(stopApp())
        })

        # done ----
        observeEvent(input$done, {
            igraph::V(rv$g)$x <- rv$xy[, 1]
            igraph::V(rv$g)$y <- rv$xy[, 2]
            if (input$text == "") {
                showNotification("Please enter a variable name", type = "warning")
            } else {
                eval(parse(text = paste0("assign(\"", input$text, "\",rv$g", ",envir = .GlobalEnv)")))
                invisible(stopApp())
            }
        })
    }


    viewer <- dialogViewer(dialogName = "Netbuilder", width = 990, height = 900)
    runGadget(ui, server, stopOnCancel = FALSE, viewer = viewer)
}


#' @export
#' @rdname Netbuilder

NetbuilderAddin <- function() {
    Netbuilder()
}


R/Netreader.R
#' Netreader
#'
#' \code{Netreader} is an RStudio-Addin that allows you to read network files. So far, only plaintext files are supported (e.g. csv,tsv,...).
#'
#' @details To run the addin, select \code{Netreader} from the Addins-menu within RStudio.
#' @return \code{Netreader} returns the created network as igraph object.
#' @import miniUI
#' @import shiny
#' @import rstudioapi
#' @importFrom igraph graph_from_adjacency_matrix graph_from_data_frame vcount vertex_attr_names set_vertex_attr get.vertex.attribute
#' @name Netreader
NULL

Netreader <- function() {
    rv <- reactiveValues(g = NULL, pathN = NULL, pathA = NULL, code = NULL)
    # ui ----
    ui <- miniPage(
        tags$head(
            tags$style(HTML("hr {border-top: 1px solid #000000;}"))
        ),
        tags$script(jscodeWidth),
        tags$script(jscodeHeight),
        tags$style(type = "text/css", ".selectize-dropdown{ width: 200px !important; }"),
        tags$style(type = "text/css", ".form-group.shiny-input-container{width:50%;}"),
        tags$style(type = "text/css", "#preview {background-color: rgba(0,0,0,0.50); color: white;}"),
        tags$style(type = "text/css", "#previewA {background-color: rgba(0,0,0,0.50); color: white;}"),
        tags$style(type = "text/css", "#netpreview {background-color: rgba(0,0,0,0.50); color: white;}"),
        tags$style(type = "text/css", "#netpreviewA {background-color: rgba(0,0,0,0.50); color: white;}"),
        tags$style(type = "text/css", "#codereview {background-color: rgba(0,0,0,1); color: white;}"),
        tags$style(type = "text/css", "#readit {background-color: rgba(30,144,255,1); color: white}"),
        tags$style(type = "text/css", "#readitA {background-color: rgba(30,144,255,1); color: white}"),
        tags$style(type = "text/css", "#netfile {background-color: rgba(30,144,255,1); color: white}"),
        tags$style(type = "text/css", "#attrfile {background-color: rgba(30,144,255,1); color: white}"),
        gadgetTitleBar("Netreader"),
        miniTabstripPanel(
            selected = "Import Network",
            miniTabPanel("Import Network",
                icon = icon("bezier-curve"),
                fillRow(
                    height = "30px", width = "100%",
                    strong("Choose network file")
                ),
                fillRow(
                    height = "50px", width = "50%",
                    # fileInput("netfile", "Choose network file")
                    actionButton("netfile", "Browse...")
                ),
                fillRow(
                    height = "50px", width = "75%",
                    verbatimTextOutput("netfilePath")
                ),
                fillRow(
                    height = line.height, width = "100%",
                    h4("File Preview (first 5 lines)")
                ),
                fillRow(
                    height = "120px", width = "100%",
                    verbatimTextOutput("preview")
                ),
                fillRow(
                    height = line.height, width = "50%",
                    checkboxInput("colnames", "Header", value = FALSE),
                    checkboxInput("rownames", "Rownames", value = FALSE),
                    checkboxInput("quotes", "Quotes", value = FALSE),
                    checkboxInput("directed", "Directed", value = FALSE)
                ),
                fillRow(
                    height = line.height, width = "50%",
                    radioButtons("readfct", "network format", choices = c("edgelist", "adjacency matrix")),
                    radioButtons("valsep", "file delimiter", choices = c("comma" = ",", "space" = " ", "tab" = "\t"))
                ),
                hr(),
                fillRow(
                    height = line.height, width = "100%",
                    actionButton("readit", "Import Network"),
                    textAreaInput("text", label = "", value = "", placeholder = "enter name", height = "35px")
                ),
                fillRow(
                    height = "120px", width = "100%",
                    verbatimTextOutput("netpreview")
                )
            ),
            # attributes ----
            miniTabPanel("Add Attributes",
                icon = icon("list-ol"),
                fillRow(
                    height = "30px", width = "100%",
                    strong("Choose attribute file")
                ),
                fillRow(
                    height = "50px", width = "100%",
                    # fileInput("attrfile", "Choose attribute file")
                    actionButton("attrfile", "Browse...")
                ),
                fillRow(
                    height = "50px", width = "75%",
                    verbatimTextOutput("attrfilePath")
                ),
                fillRow(
                    height = line.height, width = "100%",
                    h4("File Preview (first 5 lines)")
                ),
                fillRow(
                    height = "120px", width = "100%",
                    verbatimTextOutput("previewA")
                ),
                fillRow(
                    height = line.height, width = "50%",
                    checkboxInput("colnamesA", "Header", value = TRUE),
                    checkboxInput("quotesA", "Quotes", value = FALSE),
                    radioButtons("valsepA", "file delimiter", choices = c("comma" = ",", "space" = " ", "tab" = "\t"))
                ),
                tags$hr(),
                fillRow(
                    height = line.height, width = "100%",
                    actionButton("readitA", "Import Attributes")
                ),
                fillRow(
                    height = "120px", width = "100%",
                    verbatimTextOutput("netpreviewA")
                )
            ),
            # show code ----
            miniTabPanel("Review Code",
                icon = icon("code"),
                verbatimTextOutput("codereview")
            )
        )
    )
    # server ----

    server <- function(input, output, session) {
        # choose netfile path ----
        observeEvent(input$netfile, {
            rv$pathN <- file.choose()
        })
        # choose attrfile path ----
        observeEvent(input$attrfile, {
            rv$pathA <- file.choose()
        })

        output$netfilePath <- renderPrint(
            if (!is.null(rv$pathN)) {
                cat(rv$pathN)
            } else {
                cat("no file selected")
            }
        )

        output$attrfilePath <- renderPrint(
            if (!is.null(rv$pathA)) {
                cat(rv$pathA)
            } else {
                cat("no file selected")
            }
        )
        # file preview ----
        output$preview <- renderText({
            # inFile <- input$netfile
            inFile <- rv$pathN
            if (is.null(inFile)) {
                return(NULL)
            }
            txt <- readLines(inFile, n = 5)
            txt <- paste(txt, collapse = "\n")
            txt
        })

        # attribute preview ----
        output$previewA <- renderText({
            # inFile <- input$attrfile
            inFile <- rv$pathA
            if (is.null(inFile)) {
                return(NULL)
            }
            txt <- readLines(inFile, n = 5)
            txt <- paste(txt, collapse = "\n")
            txt
        })

        # network preview ----
        output$netpreview <- renderPrint({
            g <- rv$g
            if (is.null(g)) {
                return(cat("no network created yet."))
            }
            summary(g)
        })

        # network2 preview ----
        output$netpreviewA <- renderPrint({
            g <- rv$g
            if (is.null(g)) {
                return(cat("no network created yet."))
            }
            summary(g)
        })

        # codeoutput ----
        output$codereview <- renderPrint({
            cat(rv$code)
        })
        # read network ----
        observeEvent(input$readit, {
            inFile <- rv$pathN
            q <- ifelse(input$quotes, "\"", "")
            if (input$rownames) {
                A <- tryCatch(
                    utils::read.table(inFile,
                        header = input$colnames,
                        row.names = 1,
                        sep = input$valsep, quote = q,
                        stringsAsFactors = FALSE
                    ),
                    error = function(e) NULL
                )
                head <- "library(igraph)\n\n# load raw network data ----\n"
                cmd <- paste0(
                    "A <- utils::read.table(file = '", inFile, "'",
                    ",\n                header = ", input$colnames, ", row.names = 1",
                    ", sep = '", input$valsep, "'", ", quote = '", q, "', stringsAsFactors = FALSE)\n"
                )
                rv$code <- paste(head, cmd)
            } else {
                A <- tryCatch(
                    utils::read.table(inFile,
                        header = input$colnames,
                        sep = input$valsep, quote = q,
                        stringsAsFactors = FALSE
                    ),
                    error = function(e) NULL
                )
                head <- "library(igraph)\n# load raw network data ----\n"
                cmd <- paste0(
                    "A <- utils::read.table(file = '", inFile, "'",
                    ",\n                header = ", input$colnames, ", sep = '", input$valsep, "'",
                    ", quote = '", q, "', stringsAsFactors = FALSE)\n"
                )
                rv$code <- paste0(head, cmd)
            }
            if (is.null(A)) {
                showNotification("something went wrong reading the file. Check your settings", type = "error", duration = 2)
            } else {
                if (input$readfct == "adjacency matrix") {
                    mode <- ifelse(input$directed, "directed", "undirected")
                    g <- tryCatch(graph_from_adjacency_matrix(as.matrix(A), mode = mode, weighted = "weight"), error = function(e) NULL)
                    if (is.null(g)) {
                        showNotification("something went wrong creating the network.", type = "error", duration = 2)
                    } else {
                        head <- "# create network ----\n"
                        cmd <- paste0("g <- graph_from_adjacency_matrix(as.matrix(A),mode = '", mode, "')\n")
                        rv$code <- paste0(rv$code, head, cmd)
                        rv$g <- g
                        showNotification("network data successfully imported", type = "message", duration = 2)
                    }
                } else if (input$readfct == "edgelist") {
                    mode <- ifelse(input$directed, T, F)
                    g <- tryCatch(graph_from_data_frame(A, directed = mode), error = function(e) NULL)
                    if (is.null(g)) {
                        showNotification("something went wrong creating the network.", type = "error", duration = 2)
                    } else {
                        head <- "# create network ----\n"
                        cmd <- paste0("g <- graph_from_data_frame(A,directed = ", mode, ")\n")
                        rv$code <- paste0(rv$code, "\n", head, cmd)
                        rv$g <- g
                        showNotification("network data successfully imported", type = "message", duration = 2)
                    }
                }
            }
        })

        # read attributes ----
        observeEvent(input$readitA, {
            if (is.null(rv$g)) {
                showNotification("please import a network first", type = "error", duration = 2)
            } else {
                # inFile <- input$attrfile
                inFile <- rv$pathA
                q <- ifelse(input$quotesA, "\"", "")
                A <- tryCatch(
                    utils::read.table(inFile,
                        header = input$colnamesA,
                        sep = input$valsepA, quote = q,
                        stringsAsFactors = FALSE
                    ),
                    error = function(e) NULL
                )
                head <- "# load raw attribute data ----\n"
                cmd <- paste0(
                    "attrs <- utils::read.table(file = '", inFile, "'",
                    ",\n                    header = ", input$colnamesA, ", sep = '", input$valsepA, "'",
                    ", quote = '", q, "', stringsAsFactors = FALSE)\n"
                )
                rv$code <- paste0(rv$code, "\n", head, cmd)

                if (is.null(A)) {
                    showNotification("something went wrong reading the file. Check your settings", type = "error", duration = 2)
                } else {
                    if (nrow(A) != vcount(rv$g)) {
                        showNotification("The number of rows does not match the number of nodes in the network", type = "error", duration = 2)
                    } else {
                        if ("name" %in% vertex_attr_names(rv$g)) {
                            vnames <- get.vertex.attribute(rv$g, "name")
                            identCol <- which(apply(A, 2, function(x) all(x %in% vnames)))[1]
                            anames <- A[, identCol]
                            A <- A[, -identCol, drop = FALSE]
                            perm <- match(vnames, anames)
                            for (attr in names(A)) {
                                rv$g <- set_vertex_attr(rv$g, name = attr, value = A[[attr]][perm])
                            }
                            head <- "# add attributes to network ----\n"
                            cmd <- AttrNameImport
                            rv$code <- paste0(rv$code, "\n", head, cmd)
                        } else {
                            showNotification("network does not have a name attribute.\nmatching by row number instead", type = "warning", duration = 2)
                            for (attr in names(A)) {
                                rv$g <- set_vertex_attr(rv$g, name = attr, value = A[[attr]])
                            }
                            head <- "# add attributes to network ----\n"
                            cmd <- AttrRowImport
                            rv$code <- paste0(rv$code, "\n", head, cmd)
                        }
                        showNotification("Attributes successfully imported", type = "message", duration = 2)
                    }
                }
            }
        })

        # cancel ----
        observeEvent(input$cancel, {
            invisible(stopApp())
        })

        # done ----
        observeEvent(input$done, {
            if (input$text == "") {
                showNotification("Please enter a variable name", type = "warning", duration = 2)
            } else {
                eval(parse(text = paste0("assign(\"", input$text, "\",rv$g", ",envir = .GlobalEnv)")))
                invisible(stopApp())
            }
        })
    }


    viewer <- dialogViewer(dialogName = "Netreader", width = 990, height = 900)
    runGadget(ui, server, stopOnCancel = FALSE, viewer = viewer)
}


#' @export
#' @rdname Netreader

NetreaderAddin <- function() {
    Netreader()
}
